src/wallet/websocket.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import BaseWebSocket, {
+  WsOptions,
+  DEFAULT_WS_OPTIONS,
+} from '../websocket/base';
+
+export interface WalletServiceWebSocketOptions extends WsOptions {
+  walletId: string;
+};
+
+const JOIN_TIMEOUT = 5000;
+
+/**
+ * Handles websocket connections and message transmission
+ *
+ * @class
+ * @name WS
+ */
+class WalletServiceWebSocket extends BaseWebSocket {
+  // The walletId to subscribe to new events
+  private walletId: string;
+  // Timer used to detected when join wallet failed
+  private joinTimeoutTimer: ReturnType<typeof setTimeout> | null;
+
+  constructor(options: WalletServiceWebSocketOptions) {
+    const {
+      wsURL,
+      heartbeatInterval,
+      connectionTimeout,
+      retryConnectionInterval,
+      openConnectionTimeout,
+      walletId,
+    } = {
+      ...DEFAULT_WS_OPTIONS,
+      ...options,
+    };
+
+    super({
+      wsURL,
+      heartbeatInterval,
+      connectionTimeout,
+      retryConnectionInterval,
+      openConnectionTimeout,
+    });
+
+    this.walletId = walletId;
+    this.joinTimeoutTimer = null;
+  }
+
+  /**
+   * Handle message receiving from websocket
+   *
+   * @param {Object} evt Event that has data (evt.data) sent in the websocket
+   */
+  onMessage(evt) {
```

Comment:
Would it make sense to make this class `WalletServiceWebSocket` be just a wrapper around an instance of `WalletWebSocket`, instead of a sibling of it?

This way I think we would not need to re-implement this method `onMessage` here.

We would just have an instance of `WalletWebSocket` on `this.websocket`, for instance, then we would do:
```
    this.websocket.on('join-success', () => this.onJoinSuccess());
```

Does it make sense? I didn't really checked if this change change could bring difficulties in some other part of the code.

Reply:
Both `WalletWebSocket` and `WalletServiceWebSocket` extends the base `WebSocket` class in `src/websocket/base.ts`

The idea of they being extensions of the abstract base class is that they will always share common methods (like keeping track of ping/pong and calculating uptime, etc...) but may have completely different implementations for specific methods, like the `onMessage` here (that is currently very similar)